### ISSUE TITLE
Unclear how to add a type to arguments taken by a short closure

### ISSUE THREAD

**Issue by blackwolf12333** [vimeo/psalm#11027]
It's unclear to me how to type the parameter of the short closure so that psalm doesn't error on an `ArgumentTypeCoercion`. I can guarantee that the input to the closure is always the type provided, but I couldn't find a way to tell psalm this.

Is there a way to properly type the closure here, is this supposed to be possible with `@var` like described in "Off-label usage of the @var tag" on the documentation?

https://psalm.dev/r/2a616dac12

**Comment by clvarley** [comment]
I imagine you've probably already solved this, but for those who (like me) stumbled across this while searching for an answer you can do either of the following.

### Option 1: Bind the callable directly

The cleanest option is to avoid the intermediary closure entirely and just assign the desired callback.

**Example:** https://psalm.dev/r/4893c2346c

This can be done by using the [first class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php) (from PHP 8.1 onwards) or using the [Closure::fromCallable()](https://www.php.net/manual/en/closure.fromcallable.php) method. When done this way Psalm can simply use the type definitions of the target callable and understands that the `$this->test` property is merely an alias.

### Option 2: Define the param type via comment

I can't find this explicitly documented anywhere but you can also use `@param` to document the parameters of _anonymous_ functions. Personally I find this a bit cumbersome but it gets the job done. 

**Example:** https://psalm.dev/r/5aff826a0e

From what I can tell this isn't a Psalm specific quirk but instead a result of how [ReflectionFunction::getDocComment()](https://www.php.net/manual/en/reflectionfunctionabstract.getdoccomment.php) behaves.

I did try a tidier variation of this that used `@var` to document the whole closure (`Closure(ArrayType):null`) but the same type coercion error was reported.

---

**TLDR:** Where possible always prefer using the [first class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php) and bind the function directly. Otherwise, document the type with `@param`.
